doc/developer_manual/development_environment/how-to-set-up-a-development-environment.md
# How to Set Up A Development Environment for Zammad

To be able to start developing some fancy features or to make
Zammad even greater by fixing some issues, you'll need a development environment.

The following software/tools are needed for this.

Right now, we only have instructions for macOS users and Linux users using an dpkg/apt package manager
based distribution. Users of Linux distributions with other package managers should adapt accordingly
and are encouraged to contribute their info!

## Dependencies for Zammad

The following tools are either required or highly recommended to start hacking Zammad.

For macOS:

```screen
brew install postgresql forego imlib2 openssl@1.1 direnv geckodriver chromedriver shellcheck
```

For Linux:

```screen
sudo apt install postgresql libimlib2 libimlib2-dev openssl direnv shellcheck
```

Unfortunately there is no `forego` package / binary available for Linux. We recommend to build
it from [source](https://github.com/ddollar/forego) or alternatively use
[foreman](https://github.com/ddollar/foreman).

```screen
sudo mkdir -p /usr/local/lib/gecko
curl -L -k -s https://github.com/mozilla/geckodriver/releases/download/v0.32.0/geckodriver-v0.32.0-linux64.tar.gz -o - | sudo tar -xzf - -C /usr/local/lib/gecko/
sudo mv /usr/local/lib/gecko/geckodriver /usr/local/lib/gecko/geckodriver-0.32.0
sudo ln -sf /usr/local/lib/gecko/geckodriver-0.32.0 /usr/local/bin/geckodriver
```

```screen
sudo mkdir -p /usr/local/lib/chrome
curl -L -k -s https://chromedriver.storage.googleapis.com/109.0.5414.74/chromedriver_linux64.zip -o - | zcat - | sudo tee /usr/local/lib/chrome/chromedriver-109.0.5414.74 >/dev/null
sudo chmod +x /usr/local/lib/chrome/chromedriver-109.0.5414.74
sudo ln -sf /usr/local/lib/chrome/chromedriver-109.0.5414.74 /usr/local/bin/chromedriver
```

Beware chromedriver version has to match your installed Chrome browser version.

## RVM

To maintain different Ruby versions, we encourage the usage of [RVM](https://rvm.io/).

Attention: Please look up the Ruby version in the `Gemfile` and adapt it in the snippet.

For Linux and macOS:

```screen
curl -sSL https://get.rvm.io | bash -s stable --rails
rvm install ruby-3.4.7
rvm --default use 3.4.7
```

## NVM, Node.js and pnpm

We're using [NVM](https://github.com/nvm-sh/nvm) to manage all Node.js versions which are in use with Zammad.

Before executing the following snippet, please, make sure to look up the most recent version of NVM.

For Linux and macOS:

```screen
curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.5/install.sh | bash
nvm install node
npm install -g pnpm

# Then, in the zammad directory, install required modules:
cd </path/to/zammad-develop>
pnpm install
```

## Linting tools

To ensure a well-readable and maintainable code base, we're using linting tools like:

- [CoffeeLint](http://www.coffeelint.org/)
- [Stylelint](https://stylelint.io/)
- [ESLint](https://eslint.org/)
- [Markdownlint](https://github.com/DavidAnson/markdownlint)

There is also a dependency on [Docker](https://www.docker.com/) for some linting tasks, make sure it's available on your
system.

For Linux and macOS:

```screen
npm install -g coffeelint@1 stylelint
```

## Elasticsearch

Proper operation of Zammad requires [Elasticsearch](https://www.elastic.co/de/elasticsearch/).

For macOS:

```screen
brew tap elastic/tap
brew install elastic/tap/elasticsearch-full
elasticsearch-plugin install ingest-attachment
brew services start elastic/tap/elasticsearch-full
```

For Linux:

```screen
sudo apt install apt-transport-https
echo "deb [signed-by=/usr/share/keyrings/elasticsearch-keyring.gpg] https://artifacts.elastic.co/packages/8.x/apt stable main" | sudo tee /etc/apt/sources.list.d/elasticsearch.list
wget -qO - https://artifacts.elastic.co/GPG-KEY-elasticsearch | sudo gpg --dearmor -o /usr/share/keyrings/elasticsearch-keyring.gpg
sudo apt update
sudo apt install elasticsearch
sudo /usr/share/elasticsearch/bin/elasticsearch-plugin install ingest-attachment
sudo systemctl restart elasticsearch.service
```

## Redis

For macOS:

```screen
brew install redis
brew services start redis
```

For Linux (Ubuntu/Debian):

```screen
sudo apt install lsb-release

curl -fsSL https://packages.redis.io/gpg | sudo gpg --dearmor -o /usr/share/keyrings/redis-archive-keyring.gpg

echo "deb [signed-by=/usr/share/keyrings/redis-archive-keyring.gpg] https://packages.redis.io/deb $(lsb_release -cs) main" | sudo tee /etc/apt/sources.list.d/redis.list

sudo apt-get update
sudo apt-get install redis
```

Most major Linux distributions provide packages for Redis.

## Ruby

All Ruby dependencies (including development dependencies) can be installed easily via

For Linux and macOS:

```screen
cd </path/to/zammad-develop/>
bundle install
```

## Using HTTPS

Zammad uses the gem `localhost` to automatically generate self-signed certificates. This will place
`~/.local/state/localhost.rb/localhost.crt` and `~/.local/state/localhost.rb/localhost.key` files if needed. Then you
can use one of the following commands to start the development server:

```sh
VITE_RUBY_HOST=0.0.0.0 VITE_RUBY_HTTPS=true RAILS_ENV=development forego start -r -f Procfile.dev-https
# or
pnpm dev:https
```

The application will be listening on [https://localhost:3000](https://localhost:3000).

### Self-signed Certificate Exemption

By default, the browser will not allow you to access an HTTPS site with a self-signed certificate. You will need to add
an exemption by clicking on **Advanced** and choosing **Proceed (unsafe)** or **Accept the Risk and Continue**.

In Firefox, you will also have to add an exemption for WebSocket addresses, since they use a different port. Visit
[https://localhost:6042](https://localhost:6042) and [https://localhost:3036](https://localhost:3036) to kick-start the
process and then try to reload the app.

### Signed Certificate Issued via Let's Encrypt

Sometimes, using self-signed certificates might not be enough, due to some platforms still not executing the app in a
secure context. You can issue a proper signed certificate via [Let's Encrypt](https://letsencrypt.org/) service for
free. As a pre-requisite, you will need an access to a DNS table of a custom domain and a local instance of
[Docker](https://www.docker.com/).

First, decide on a subdomain for your app, i.e. if you own `example.com` you may want to use `localhost.example.com`.

Next, run the following Docker container to start the DNS01 challenge process to verify you own the domain in question:

```sh
docker run --rm -it -v /path/to/certs:/etc/letsencrypt certbot/dns-cloudflare certonly --manual --preferred-challenges dns --email you@example.com --agree-tos --no-eff-email --key-type rsa -d localhost.example.com
```

Where:

- `/path/to/certs` is a local directory where your certificate files will be stored
- `you@example.com` is your email address
- `localhost.example.com` is the FQDN of your subdomain

When asked to deploy a DNS TXT record by certbot, open the DNS table of your domain. Add a TXT record with suggested
name in form of `_acme-challenge.localhost.example.com.` and suggested random value. Save the record and wait some
seconds for changes to propagate (this may depend on your DNS host).

Then, press Enter to continue the challenge process. If the certbot identifies your DNS record, it will automatically
issue an appropriate certificate. Do not proceed if there was an error logged, resolve it first.

Next, backup your current self-signed certificate files (if they exist) and create symbolic links to the new ones:

```sh
cd ~/.local/state/localhost.rb
mv localhost.crt localhost.crt.self-signed
mv localhost.key localhost.key.self-signed
ln -s /path/to/certs/live/localhost.example.com/cert.pem ~/.local/state/localhost.rb/localhost.crt
ln -s /path/to/certs/live/localhost.example.com/privkey.pem ~/.local/state/localhost.rb/localhost.key
```

You may need to adjust the paths depending on your subdomain name.

Next, add an A DNS record for your subdomain that points to your local IP. You can find out your local IP via
`ifconfig` or a similar command.

For example, if your local IP is `192.168.0.39` and your subdomain is `localhost.example.com`, add an A DNS record with
the name of `localhost` and point it to `192.168.0.39`. This will allow you to access the app from within your local
network only by using the proper FQDN: perfect for testing the app on mobile devices.

Finally, start the development server with `pnpm dev:https` command. You can now access the app via
[https://localhost.example.com:3000](https://localhost.example.com:3000) and it should show up as a trusted site.


doc/developer_manual/development_environment/how-to-set-up-a-development-environment-using-docker.md
# How to Set Up A Development Environment for Zammad using docker & docker compose

## Run `bin/docker-setup` shortcut at first start

It will:

- Create blank .env file
- Build docker image
- Install javascript and ruby dependencies into volumes

## Then run `docker compose up` to start application


CHANGES.md
# Changes

## Zammad 1.1.0
 - 2016-11-04 Implemented issue#[349](https://github.com/zammad/zammad/issues/349) - ticket search cache issue.
 - 2016-10-28 Implemented issue#[317](https://github.com/zammad/zammad/issues/317) - Display origin url from tweet/facebook post in article view.

## Zammad 1.0.1
 - 2016-11-04 Fixed issue#[312](https://github.com/zammad/zammad/issues/312) - In object manager, blank line in object deleted all values for selections.
 - 2016-11-04 Fixed issue#[286](https://github.com/zammad/zammad/issues/286) - Answers from google mail shows complete quote.
 - 2016-11-03 Fixed issue#[348](https://github.com/zammad/zammad/issues/348) - IMAP mail fetching stops because of broken spam email (invalid Content-Transfer-Encoding header)
 - 2016-10-29 Fixed issue#[326](https://github.com/zammad/zammad/issues/326) - Create emails via fetchmail / procmail.
 - 2016-10-28 Fixed issue#[251](https://github.com/zammad/zammad/issues/251) - Page posts from facebook and tweets from twitter as tickets (new tickets from own posts / tweets will be created as closed now).
 - 2016-10-28 Fixed issue#[323](https://github.com/zammad/zammad/issues/323) - Unable to create tickets if SLA with condition article.subject is set.
 - 2016-10-27 Fixed issue#[268](https://github.com/zammad/zammad/issues/268) - No space at comment on ticket bulk action.
 - 2016-10-27 Fixed issue#[313](https://github.com/zammad/zammad/issues/313) - Admin -> Channels -> Email -> Filters not shown / no name shown.
 - 2016-10-23 Fixed issue#[277](https://github.com/zammad/zammad/issues/277) - Chat history wrong on app/page reload.


app/frontend/shared/types/config.ts
export interface ConfigList {
  api_path: string
  'active_storage.web_image_content_types': string[]
  'auth_saml_credentials.display_name'?: string
  'auth_openid_connect_credentials.display_name'?: string
  ai_assistance_text_tools: boolean
  ai_assistance_text_tools_fixed_instructions: string
  ai_assistance_ticket_summary: boolean
  ai_assistance_ticket_summary_config: unknown
  ai_provider: boolean
  api_password_access?: boolean | null
  api_token_access?: boolean | null
  auth_facebook?: boolean | null
  auth_github?: boolean | null
  auth_gitlab?: boolean | null
  auth_google_oauth2?: boolean | null
  auth_linkedin?: boolean | null
  auth_microsoft_office365?: boolean | null
  auth_openid_connect?: boolean | null
  auth_saml?: boolean | null
  auth_sso?: boolean | null
  auth_twitter?: boolean | null
  auth_weibo?: boolean | null
  chat?: boolean | null
  chat_agent_idle_timeout: string
  checklist?: boolean | null
  core_workflow_ajax_mode?: boolean | null
  cti_integration?: boolean | null
  customer_ticket_create?: boolean | null
  customer_ticket_create_group_ids: unknown
  datepicker_show_calendar_weeks?: boolean | null
  default_controller: string
  defaults_calendar_subscriptions_tickets: unknown
  developer_mode: boolean
  exchange_integration?: boolean | null
  fqdn: string
  github_integration?: boolean | null
  gitlab_integration?: boolean | null
  http_type?: 'https' | 'http' | null
  idoit_integration?: boolean | null
  import_backend: string
  import_mode?: boolean | null
  kb_active: boolean
  kb_active_publicly: boolean
  kb_multi_lingual_support: boolean
  ldap_integration?: boolean | null
  locale_default: string
  maintenance_login: boolean
  maintenance_login_message: string
  maintenance_mode: boolean
  organization: string
  password_max_login_failed?: 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 13 | 14 | 15 | 16 | 17 | 18 | 19 | 20 | null
  pgp_config: unknown
  pgp_integration?: boolean | null
  pgp_recipient_alias_configuration: boolean
  placetel_integration?: boolean | null
  pretty_date_format: 'relative' | 'absolute' | 'timestamp'
  product_logo: string
  product_name: string
  session_timeout: unknown
  sipgate_integration?: boolean | null
  smime_config: unknown
  smime_integration?: boolean | null
  system_id?: 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 | 20 | 21 | 22 | 23 | 24 | 25 | 26 | 27 | 28 | 29 | 30 | 31 | 32 | 33 | 34 | 35 | 36 | 37 | 38 | 39 | 40 | 41 | 42 | 43 | 44 | 45 | 46 | 47 | 48 | 49 | 50 | 51 | 52 | 53 | 54 | 55 | 56 | 57 | 58 | 59 | 60 | 61 | 62 | 63 | 64 | 65 | 66 | 67 | 68 | 69 | 70 | 71 | 72 | 73 | 74 | 75 | 76 | 77 | 78 | 79 | 80 | 81 | 82 | 83 | 84 | 85 | 86 | 87 | 88 | 89 | 90 | 91 | 92 | 93 | 94 | 95 | 96 | 97 | 98 | 99 | null
  system_init_done: boolean
  system_online_service: boolean
  tag_new?: boolean | null
  ticket_agent_default_notifications: unknown
  ticket_allow_expert_conditions?: boolean | null
  ticket_auto_assignment?: boolean | null
  ticket_auto_assignment_selector: unknown
  ticket_auto_assignment_user_ids_ignore: unknown
  ticket_conditions_allow_regular_expression_operators?: boolean | null
  ticket_define_email_from?: 'SystemAddressName' | 'AgentNameSystemAddressName' | 'AgentName' | null
  ticket_define_email_from_separator: string
  ticket_duplicate_detection?: boolean | null
  ticket_duplicate_detection_attributes: unknown
  ticket_duplicate_detection_body: string
  ticket_duplicate_detection_permission_level: string
  ticket_duplicate_detection_role_ids: unknown
  ticket_duplicate_detection_search: string
  ticket_duplicate_detection_show_tickets: boolean
  ticket_duplicate_detection_title: string
  ticket_hook: string
  ticket_organization_reassignment: boolean
  ticket_secondary_action: string
  time_accounting?: boolean | null
  time_accounting_selector: unknown
  time_accounting_type_default: string
  time_accounting_types: boolean
  time_accounting_unit: string
  time_accounting_unit_custom: string
  timezone_default: string
  two_factor_authentication_enforce_role_ids: unknown
  two_factor_authentication_method_authenticator_app?: boolean | null
  two_factor_authentication_method_security_keys?: boolean | null
  two_factor_authentication_recovery_codes?: boolean | null
  ui_desktop_beta_switch: boolean
  ui_richtext_bubble_menu?: boolean | null
  ui_sidebar_open_ticket_indicator_colored?: boolean | null
  ui_table_group_by_show_count?: boolean | null
  ui_task_mananger_max_task_count: number
  ui_ticket_add_article_hint: unknown
  ui_ticket_create_available_types: ('phone-in' | 'phone-out' | 'email-out')[]
  ui_ticket_create_default_type: 'phone-in' | 'phone-out' | 'email-out'
  ui_ticket_create_notes: unknown
  ui_ticket_overview_query_polling: unknown
  ui_ticket_overview_ticket_limit: number
  ui_ticket_priority_icons?: boolean | null
  ui_ticket_zoom_article_delete_timeframe: number
  ui_ticket_zoom_article_email_full_quote?: boolean | null
  ui_ticket_zoom_article_email_full_quote_header?: boolean | null
  ui_ticket_zoom_article_email_subject?: boolean | null
  ui_ticket_zoom_article_note_new_internal?: boolean | null
  ui_ticket_zoom_article_visibility_confirmation_dialog?: boolean | null
  ui_ticket_zoom_attachments_preview?: boolean | null
  ui_ticket_zoom_sidebar_article_attachments?: boolean | null
  ui_user_organization_selector_with_email?: boolean | null
  user_create_account?: boolean | null
  user_lost_password?: boolean | null
  user_show_password_login?: boolean | null
  websocket_backend: string
  websocket_port: string
  [key: string]: unknown
}
